### PR TITLE
Enhance gnmi_cli_py

### PR DESCRIPTION
**Pull request by ganglyu**
1. Upgrade grpcio and grpcio-tools
2. Support origin in gnmi prefix
3. Print grpc error code
4. Support PROTO encoding for gnmi get
5. Ignore / in []
6. Support proto_bytes for gnmi set
7. Support multiple path and value for gnmi set, get and subscribe

Upstream PR for google/gnxi:
https://github.com/google/gnxi/pull/350
https://github.com/google/gnxi/pull/351

### CODE REVIEW

**Comment by qiluo-msft** [comment]
>         print("Client receives an exception '{}' indicating gNMI server is shut down and Exiting ..."

Original error message seems useful. If you do not check err.code, will you print similar fancy message? If not, let's use `if-else` to print UNAVAILABLE message and general message. #Closed

---
Refers to: gnmi_cli_py/py_gnmicli.py:557 in 31ba981. [](commit_id = 31ba981cb5bc2abac705b98e459f2126eb0aaabe, deletion_comment = True)

File: gnmi_cli_py/py_gnmicli.py [hunk 1]
```diff
@@ -156,7 +158,7 @@ def _create_parser():
                       help='subscription mode [0=TARGET_DEFINED, 1=ON_CHANGE, 2=SAMPLE]')
   parser.add_argument('--update_count', default=0, type=int, help='Max number of streaming updates to receive. 0 means no limit.')
   parser.add_argument('--subscribe_mode', default=0, type=int, help='[0=STREAM, 1=ONCE, 2=POLL]')
-  parser.add_argument('--encoding', default=0, type=int, help='[0=JSON, 1=BYTES, 2=PROTO, 3=ASCII, 4=JSON_IETF]')
+  parser.add_argument('--encoding', default=4, type=int, help='[0=JSON, 1=BYTES, 2=PROTO, 3=ASCII, 4=JSON_IETF]')
```
Comment (qiluo-msft) [inline gnmi_cli_py/py_gnmicli.py]: >4 [](http://example.com/codeflow?start=44&length=1)

This is a valid bugfix on its own. Suggest separate into a standalone commit, not to mix with proto feature. #Closed

File: gnmi_cli_py/py_gnmicli.py [hunk 2]
```diff
@@ -515,11 +531,18 @@ def main():
...
               ' with the following gNMI Path\n', '-'*25, '\n', paths)
-        response = _get(stub, paths, user, password, prefix)
+        response = _get(stub, paths, user, password, prefix, encoding)
         print('The GetResponse is below\n' + '-'*25 + '\n')
-        if form == 'protobuff':
+        if encoding == 'PROTO':
+          i = 0
+          for notification in response.notification:
+            for update in notification.update:
+              with open(proto_list[i], 'wb') as fp:
```
Comment (qiluo-msft) [inline gnmi_cli_py/py_gnmicli.py:599]: >[i] [](http://example.com/codeflow?start=34&length=3)

Possible to index out of range? #Closed

File: gnmi_cli_py/py_gnmicli.py [hunk 3]
```diff
@@ -113,6 +113,8 @@ def _create_parser():
                       '\nCan be Leaf value or JSON file. If JSON file, prepend'
                       ' with "@"; eg "@interfaces.json".',
                       required=False)
+  parser.add_argument('--proto', type=str, help='Output files for proto bytes',
+                      nargs="+", required=False)
```
Comment (qiluo-msft) [inline gnmi_cli_py/py_gnmicli.py:116]: >"+" [](http://example.com/codeflow?start=28&length=3)

You assume at least one file. If the query is expected to fail, do you allow gnmicli to have 0 args? #Closed

**Comment by qiluo-msft** [comment]
> def _path_names(xpath):

Could you check the go implementation [https://github.com/google/gnxi/blob/2cca83d741c326d88cc3652cfbac26161bbb7209/utils/xpath/xpath.go#L255](https://github.com/google/gnxi/blob/2cca83d741c326d88cc3652cfbac26161bbb7209/utils/xpath/xpath.go#L255).
Is it more robust? #Closed

---
Refers to: gnmi_cli_py/py_gnmicli.py:173 in 88adc85. [](commit_id = 88adc85db00d96c5190b1bd61fd410b914159ff0, deletion_comment = False)

File: gnmi_cli_py/py_gnmicli.py [hunk 4]
```diff
@@ -275,6 +300,13 @@ def _get_val(json_value):
       raise JsonReadError('Error while loading JSON: %s' % str(e))
     val.json_ietf_val = json.dumps(set_json).encode()
     return val
+  elif '$' in json_value:
+    try:
+      proto_bytes = six.moves.builtins.open(json_value.strip('$'), 'rb').read()
+    except (IOError, ValueError) as e:
+      raise Exception('Error while loading proto: %s' % str(e))
```
Comment (qiluo-msft) [inline gnmi_cli_py/py_gnmicli.py]: >Exception [](http://example.com/codeflow?start=12&length=9)

Suggest to use `ValueError`. Does the error message contains the filename? If not, suggest add it to help troubleshoot. #Closed

File: gnmi_cli_py/py_gnmicli.py [hunk 5]
```diff
@@ -275,6 +306,15 @@ def _get_val(json_value):
...
+    val.proto_bytes = proto_bytes
+    return val
+  elif '#' in json_value:
```
Comment (qiluo-msft) [inline gnmi_cli_py/py_gnmicli.py]: >'#' in json_value [](http://example.com/codeflow?start=7&length=17)

Is it a new feature in command line? I do not see any usage text or code comment, so hard to read. #Closed

File: gnmi_cli_py/py_gnmicli.py [hunk 6]
```diff
@@ -113,6 +113,12 @@ def _create_parser():
                       '\nCan be Leaf value or JSON file. If JSON file, prepend'
                       ' with "@"; eg "@interfaces.json".',
                       required=False)
+  parser.add_argument('--value_list', type=str, help='Value for SetRequest.'
```
Comment (qiluo-msft) [inline gnmi_cli_py/py_gnmicli.py]: >value_list [](http://example.com/codeflow?start=25&length=10)

Possible to reuse original `-val` with `nargs="+"` ? #Closed

File: gnmi_cli_py/py_gnmicli.py [hunk 7]
```diff
@@ -127,7 +133,9 @@ def _create_parser():
                       'Target when establishing secure gRPC channel.',
                       required=False, action='store_true')
   parser.add_argument('-x', '--xpath', type=str, help='The gNMI path utilized'
-                      'in the GetRequest or Subscirbe', required=True)
+                      'in the GetRequest or Subscirbe')
+  parser.add_argument('--xpath_list', type=str, help='The gNMI path utilized'
```
Comment (qiluo-msft) [inline gnmi_cli_py/py_gnmicli.py]: >--xpath_list [](http://example.com/codeflow?start=23&length=12)

Possible to reuse original `-x` with `nargs="+"` ? #Closed

File: gnmi_cli_py/py_gnmicli.py [hunk 8]
```diff
@@ -197,6 +197,11 @@ def _parse_path(p_names):
     XpathError: Unabled to parse the xpath provided.
   """
   gnmi_elems = []
+  gnmi_origin = None
+  if p_names and ":" in p_names[0]:
+    res = p_names[0].split(":", 1)
```
Comment (qiluo-msft) [inline gnmi_cli_py/py_gnmicli.py]: >":" [](http://example.com/codeflow?start=27&length=3)

If it your invention or gnmi protocol? #Closed

Comment (ganglyu) [reply gnmi_cli_py/py_gnmicli.py]: This is not gnmi protocol.
Google gnxi was using the same solution: https://github.com/google/gnxi/commit/89f51f0ce1e24323174626e8d236623b2d12f09b
Now they use prefix to support origin:
https://github.com/google/gnxi/pull/295
Please let me know which one is better.

File: gnmi_cli_py/py_gnmicli.py (hunk 7 above)
Comment (ganglyu) [reply gnmi_cli_py/py_gnmicli.py]: Updated.

File: gnmi_cli_py/py_gnmicli.py (hunk 6 above)
Comment (ganglyu) [reply gnmi_cli_py/py_gnmicli.py]: updated

File: gnmi_cli_py/py_gnmicli.py (hunk 5 above)
Comment (ganglyu) [reply gnmi_cli_py/py_gnmicli.py]: I have added comment.

File: gnmi_cli_py/py_gnmicli.py (hunk 2 above)
Comment (ganglyu) [reply gnmi_cli_py/py_gnmicli.py:599]: Updated.

File: gnmi_cli_py/py_gnmicli.py (hunk 4 above)
Comment (ganglyu) [reply gnmi_cli_py/py_gnmicli.py]: Updated.

Comment (ganglyu) [reply gnmi_cli_py/py_gnmicli.py]: updated.

File: gnmi_cli_py/py_gnmicli.py (hunk 3 above)
Comment (ganglyu) [reply gnmi_cli_py/py_gnmicli.py:116]: updated to use *

File: gnmi_cli_py/py_gnmicli.py (hunk 1 above)
Comment (ganglyu) [reply gnmi_cli_py/py_gnmicli.py]: This change is not necessary, so I have reverted it.

File: gnmi_cli_py/py_gnmicli.py (hunk 5 above)
Comment (qiluo-msft) [reply gnmi_cli_py/py_gnmicli.py]: "Use '#' to indicate empty value" is kind of hacky to me.
could you just use empty python string to "to indicate empty value"?

Comment (ganglyu) [reply gnmi_cli_py/py_gnmicli.py]: Updated

File: gnmi_cli_py/py_gnmicli.py (hunk 8 above)
Comment (ganglyu) [reply gnmi_cli_py/py_gnmicli.py]: According to gnmi spec: In the case that a prefix is specified, it MUST specify any required origin.
Now I use prefix to support origin

**Review by qiluo-msft — APPROVED** [review]

**Review by qiluo-msft — APPROVED** [review]
